core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestRemoveEmptyUnionBranches.java

```diff
@@ -0,0 +1,108 @@
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableListMultimap;
+import com.google.common.collect.ImmutableMap;
+import io.trino.sql.planner.Symbol;
+import io.trino.sql.planner.iterative.rule.test.BaseRuleTest;
+import io.trino.sql.tree.NullLiteral;
+import org.testng.annotations.Test;
+
+import java.util.List;
+
+import static io.trino.sql.planner.assertions.PlanMatchPattern.expression;
+import static io.trino.sql.planner.assertions.PlanMatchPattern.project;
+import static io.trino.sql.planner.assertions.PlanMatchPattern.union;
+import static io.trino.sql.planner.assertions.PlanMatchPattern.values;
+
+public class TestRemoveEmptyUnionBranches
+        extends BaseRuleTest
+{
+    @Test
+    public void testDoNotEliminateNonEmptyBranches()
+    {
+        tester().assertThat(new RemoveEmptyUnionBranches())
+                .on(p -> {
+                    Symbol output = p.symbol("output");
+                    Symbol input1 = p.symbol("input1");
+                    Symbol input2 = p.symbol("input2");
+
+                    return p.union(
+                            ImmutableListMultimap.<Symbol, Symbol>builder()
+                                    .put(output, input1)
+                                    .put(output, input2)
+                                    .build(),
+                            ImmutableList.of(
+                                    p.values(1, input1),
+                                    p.values(2, input2)));
+                })
+                .doesNotFire();
+    }
+
+    @Test
+    public void testRemoveEmptyBranches()
+    {
+        tester().assertThat(new RemoveEmptyUnionBranches())
+                .on(p -> {
+                    Symbol output = p.symbol("output");
+                    Symbol input1 = p.symbol("input1");
+                    Symbol input2 = p.symbol("input2");
+                    Symbol input3 = p.symbol("input3");
+                    Symbol input4 = p.symbol("input4");
+
+                    return p.union(
+                            ImmutableListMultimap.<Symbol, Symbol>builder()
+                                    .put(output, input1)
+                                    .put(output, input2)
+                                    .put(output, input3)
+                                    .put(output, input4)
+                                    .build(),
+                            ImmutableList.of(
+                                    p.values(1, input1),
+                                    p.values(0, input2),
+                                    p.values(2, input3),
+                                    p.values(0, input4)));
+                })
+                .matches(
+                        union(
+                                values(List.of("input1"), List.of(List.of(new NullLiteral()))),
+                                values(List.of("input2"), List.of(List.of(new NullLiteral()), List.of(new NullLiteral())))));
```

Comment:
Technically, it could be anything. These names are supposed to be matcher aliases and not necessarily related to the symbols used in the plan. I'll fix it for consistency, though.